### DISCUSSION TITLE
How can I access the tcp seq number in the struct sk_buff structure

### DISCUSSION THREAD

**Discussion by miaochang123** [bpftrace/bpftrace#2435] [Q&A]
I want to check the  tcp packet seq number of  the  `__tcp_transmit_skb`'s second parameter

```
kfunc:__tcp_transmit_skb
    struct sock * sk
    struct sk_buff * skb
    int clone_it
    gfp_t gfp_mask
    u32 rcv_nxt
    int retval
```

in kernel  we access it by the  Macro `TCP_SKB_CB(skb)->seq`
```
#define TCP_SKB_CB(__skb)	((struct tcp_skb_cb *)&((__skb)->cb[0]))  
```
but bpftrace seems do not allow me to do the  Type conversion

```
$ bpftrace -e 'kprobe:__tcp_transmit_skb {
  printf("seq = %u", (((struct tcp_skb_cb *)((struct sk_buff*)arg1->cb))->seq))
}'

stdin:1:27-97: ERROR: Can not access field 'cb' on expression of type 'unsigned int64'
kprobe:__tcp_transmit_skb {printf("seq = %u",(((struct tcp_skb_cb *)((struct sk_buff*)arg1->cb))->seq)) }
```

**Comment by fbs** [comment]
` 'cb' on expression of type 'unsigned int64'`

my guess is that its doing the arg1 dereference before the cast.  

To avoid parenthesishell i'd do something like:

```
$ bpftrace -e 'kprobe:__tcp_transmit_skb {
  $skbuf = (struct sk_buff*) arg1;
  $tcp = (struct tcp_skb_cb *) $skbuf->cb;
  printf("seq = %u", $tcp->seq);
}'
```

**Reply by burchell1992** [reply]
appreciate your help very much， 
But according to my test ， this script  can be compiled through 
but the obtained seq numbers are all 0, 
It doesn't look right

[root@nj-k8s-01 ~]# bpftrace -e 'kprobe:__tcp_transmit_skb {
  $skbuf = (struct sk_buff*) arg1;
  $tcp = (struct tcp_skb_cb *) $skbuf->cb;
  printf("seq = %u", $tcp->seq);
}'
Attaching 1 probe...
seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq = 0seq =
